platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -116,35 +114,66 @@ public class HalfDiskHashMap implements AutoCloseable, Snapshotable, FileStatist
+
+    /**
+     * Number of bucket tasks that can be scheduled at the moment, i.e. MAX_IN_FLIGHT minus
+     * the number of buckets currently being processed
+     */
+    private final AtomicInteger bucketPermits = new AtomicInteger(MAX_IN_FLIGHT);
+
+    /**
+     * The last created "store bucket" task, or null if no such tasks have been created yet in
+     * the current flush. This reference is used to set store task dependencies to make sure
+     * only a single bucket is written to disk at a time
+     */
+    private final AtomicReference<StoreBucketTask> lastStoreTask = new AtomicReference<>();
+
+    /**
+     * Number of "store bucket" tasks created so far in the current flush. This counter is
+     * compared against {@link #numBuckets} to manage the first and the last "store bucket"
+     * task dependencies
+     */
+    private final AtomicInteger storeBucketTasksCreated = new AtomicInteger();
+
+    /**
+     * A task used in {@link #endWriting()} to wait for all submitted tasks to complete. It
+     * is set as a "next" dependency for the last created "store bucket" task.
+     */
+    private final AtomicReference<AbstractTask> notifyTaskRef = new AtomicReference<>();
+
+    /** A holder for the first exception occured during endWriting() tasks */
+    private final AtomicReference<Exception> exceptionOccurred = new AtomicReference<>();
+
+    /** Fork-join pool for HDHM.endWriting() */
+    private static volatile ForkJoinPool flushingPool = null;
 
-    private static ExecutorService getFlushExecutor() {
-        ExecutorService exec = flushExecutor;
-        if (exec == null) {
+    private static ForkJoinPool getFlushingPool() {
+        ForkJoinPool pool = flushingPool;
+        if (pool == null) {
             synchronized (HalfDiskHashMap.class) {
-                exec = flushExecutor;
-                if (exec == null) {
-                    final MerkleDbConfig config = ConfigurationHolder.getConfigData(MerkleDbConfig.class);
-                    exec = Executors.newFixedThreadPool(
-                            config.getNumHalfDiskHashMapFlushThreads(),
-                            new ThreadConfiguration(getStaticThreadManager())
-                                    .setComponent(MERKLEDB_COMPONENT)
-                                    .setThreadName("HalfDiskHashMap Flushing")
-                                    .setExceptionHandler((t, ex) -> logger.error(
-                                            EXCEPTION.getMarker(), "Uncaught exception during HDHM flushing", ex))
-                                    .buildFactory());
-                    flushExecutor = exec;
+                pool = flushingPool;
+                if (pool == null) {
+                    final MerkleDbConfig vmConfig = ConfigurationHolder.getConfigData(MerkleDbConfig.class);
+                    final int hashingThreadCount = vmConfig.getNumHalfDiskHashMapFlushThreads();
+                    pool = new ForkJoinPool(hashingThreadCount);
```

Comment:
There is a subtle difference between the proposed configuration and the previous one: the calling thread in `endWriting()` participated in processing by performing writes in a busy loop; now, it's waiting in `join()`.
I suggest increasing `merkleDb.percentHalfDiskHashMapFlushThreads` from 50% to 75% to compensate for that (actually, improve on that). `VirtualMapBench` shows 5% improvement.